webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,219 @@
+    if (this.#shouldSkip(newSub) && this.#syncSent) {
+      this.#advance();
+      return;
+    }
+
+    this.#numSyncs += 1;
+    const sync_id = this.#numSyncs.toString();
+    const spec = reduce(
+      newSub,
+      (spec, ent, k) => {
+        spec['known'][k] = ent.keyCache.known();
+        spec['subscribe'][k] = {
+          ...ent.spec.toWire(),
+          since: ent.keyCache.maxSeq(),
+        };
+        return spec;
+      },
+      { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, StreamContent>,
+    );
+
+    this.#curSub = newSub;
+    this.#ws && this.#ws.send(JSON.stringify(spec));
+    this.#syncSent = sync_id;
+  }
+
+  #processMsg(): void {
+    while (this.#pendingMsg.length > 0) {
+      const msg = this.#pendingMsg.shift();
+      if (!msg) break;
+      if (msg['sync_id']) {
+        if (!msg['complete']) {
+          this.#syncStarted = msg['sync_id'];
+          this.#isLoading?.(false);
+        } else {
+          this.#syncComplete = msg['sync_id'];
+          this.#isLoading?.(true);
+        }
+      } else if (this.#syncSent === this.#syncStarted) {
+        // Ignore all messages between when we send a new subscription and when the
+        // sync-start message for that subscription arrives.  These are the online
+        // updates for a subscription we no longer care about.
+
+        forEach(msg, (val, k) => {
+          if (k.includes('_deleted')) {
+            const stream_key = trimEnd(k, '_deleted') as Streamable;
+            const deleted_keys = decode_keys(val as string);
+
+            this.#curSub?.[stream_key].keyCache.delete_msg(deleted_keys);
+            this.#onDelete(stream_key, deleted_keys);
+          } else {
+            this.#curSub?.[StreamEntityMap[k]].keyCache.upsert([val.id], val.seq);
+            this.#onUpsert(msg);
+          }
+        });
+      }
+    }
+  }
+
+  #processSubscription() {
+    if (!this.#curSub && (!this.#subs || this.#subs.length === 0)) return;
+
+    let spec: Record<Streamable, Subscription> | undefined;
+    // Just connected/reconnected
+    if (!this.#syncSent) {
+      // Resend current subscription if current sync not completed
+      if (this.#curSub && (!this.#syncComplete || this.#syncStarted !== this.#syncComplete)) {
+        spec = this.#curSub;
+      } else {
+        spec = this.#subs.shift();
+      }
+      this.#sendSpec(spec);
+    } else if (this.#subs && this.#syncComplete === this.#syncSent) {
+      // for established connection, only send a new sub when current sub is completed
+      spec = this.#subs.shift();
+      this.#sendSpec(spec);
+    }
+  }
+
+  #advance(): void {
+    // We want to shut down
+    if (this.#closedByClient) {
+      if (this.#ws && this.#ws.readyState !== this.#ws.CLOSED) {
+        if (this.#ws.readyState !== this.#ws.CLOSING) {
+          this.#ws.close();
+        }
+      }
+      return;
+    }
+
+    // We got shut down and we wait till timeout finish to reconnect
```

Comment:
again, comment placement is wrong, also the phrasing of "we got shut down" reads like the caller shut us down.  Say "our websocket broke" or something to clearly distinguish it from the case where the caller actually shut us down.